buildpacks/ruby/src/layers/bundle_download_layer.rs

```diff
@@ -1,161 +1,163 @@
+                "Bundler version ({old} to {now})",
+                old = style::value(other.version.to_string()),
+                now = style::value(self.version.to_string())
+            ));
         }
+        differences
     }
+}
 
-    fn create(
-        &mut self,
-        _context: &BuildContext<Self::Buildpack>,
-        layer_path: &Path,
-    ) -> Result<LayerResult<Self::Metadata>, RubyBuildpackError> {
-        let bin_dir = layer_path.join("bin");
-        let gem_path = layer_path;
-
-        let mut cmd = Command::new("gem");
-        cmd.args([
-            "install",
-            "bundler",
-            "--version", // Specify exact version to install
-            &self.metadata.version.to_string(),
-        ])
-        .env_clear()
-        .envs(&self.env);
-
-        // Format `gem install --version <version>` without other content for display
-        let short_name = fun_run::display(&mut cmd);
-
-        // Arguments we don't need in the output
-        cmd.args([
-            "--install-dir", // Directory where bundler's contents will live
-            &layer_path.to_string_lossy(),
-            "--bindir", // Directory where `bundle` executable lives
-            &bin_dir.to_string_lossy(),
-            "--force",       // Overwrite if it already exists
-            "--no-document", // Don't install ri or rdoc documentation, which takes extra time
-            "--env-shebang", // Start the `bundle` executable with `#! /usr/bin/env ruby`
-        ]);
-
-        log_step_timed(format!("Running {}", fmt::command(short_name)), || {
-            cmd.named_output().map_err(|error| {
-                fun_run::map_which_problem(error, cmd.mut_cmd(), self.env.get("PATH").cloned())
-            })
-        })
-        .map_err(RubyBuildpackError::GemInstallBundlerCommandError)?;
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub(crate) struct MetadataV1 {
+    pub(crate) version: ResolvedBundlerVersion,
+}
 
-        LayerResultBuilder::new(self.metadata.clone())
-            .env(
-                LayerEnv::new()
-                    .chainable_insert(Scope::All, ModificationBehavior::Delimiter, "PATH", ":")
-                    .chainable_insert(
-                        Scope::All,
-                        ModificationBehavior::Prepend,
-                        "PATH", // Ensure this path comes before default bundler that ships with ruby, don't rely on the lifecycle
-                        bin_dir,
-                    )
-                    .chainable_insert(Scope::All, ModificationBehavior::Delimiter, "GEM_PATH", ":")
-                    .chainable_insert(
-                        Scope::All,
-                        ModificationBehavior::Prepend,
-                        "GEM_PATH", // Bundler is a gem too, allow it to be required
-                        gem_path,
-                    ),
-            )
-            .build()
-    }
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum MetadataError {
```

Comment:
So an error might happen during migration between metadata versions and you have this enum to represent that. There's currently only a single metadata version at the moment (`MetadataV1`) which implies that there are no migrations to perform so there are zero errors to represent. Therefore this empty enum exists more as a placeholder for future migrations?

Reply:
That's correct. I'm effectively laying out an empty structure that people can use later. It means that all layers that use these shared components will all be standardized. If you know how to modify one, you know how to modify all of them.

Technically I could use https://doc.rust-lang.org/std/convert/enum.Infallible.html instead. But that would require the first dev who needs to introduce a metadata migration to understand more about what's going on, enough to create a new enum and replace the error type.